Mapping collection

tdonohue opened the pull request:
Replacement for #35 (@benbosman and @KevinVdV) and #44 (@abollini). 

This PR includes alterations to the items.md contract to add for mapping of items to collections.
JIRA: https://jira.duraspace.org/browse/DS-4097

I've attempted to bring in feedback from both prior PRs. Here's what I've done:
- [x] Started from #35 
- [x] Changed all "internal" atmire.com URLs to same base URL we use elsewhere. Yes, these links won't resolve, but that's fine.  Likely we should replace all URLs in the future, and we just want consistency for now
- [x] Added `PUT` definition from #44  (was missing from #35).

**BEFORE WE CAN MERGE THIS**
One outstanding question remains...we have a discrepancy on the `DELETE` method format:
* #35 expects DELETE to use "text/uri-list" (like PUT and POST)
* #44 expects DELETE to use `items/[:uuid]/mappedCollections/[:uuid_of_collection_to_remove]`

In the Spring Data REST community, they seem to recommend using `PUT` for deletion (by calling `PUT` with an empty body). See: https://stackoverflow.com/a/28713232/3750035  Though, there's also an answer more like #44 here: https://stackoverflow.com/a/40785090/3750035

For ease of parsing all these endpoints the same way, I wonder if we should simply use `PUT` for deletion.

benbosman commented on items.md:
```diff
@@ -245,14 +246,227 @@ Status codes:
+ * 405: if the item is a template item
+ * 422: if the specified collection is not found or is the owningCollection of the item
+
+**PUT /api/core/items/<item:uuid>/mappedCollections**
```
At the moment there's no use case for a `PUT`, so that's not planned to be implemented
To avoid any confusion, I assume it's best to leave it out here

tdonohue commented:
@benbosman and @abollini : I've updated this REST Contract PR based on your feedback:

- [x] Per feedback from @benbosman: I've noted that `PUT` is not supported at this time, but left in the section (to remain consistent with other areas of the Contract)
- [x] Per feedback from @abollini : I've removed the detailed example GET response.  Attempting to keep these detailed responses up-to-date will be very difficult until the Contract is closer to being finalized.  If we decide we want detailed response examples for all GET requests, we can add them in later. 
- [x] Per feedback from @abollini : I've updated the `DELETE` request format to pass the UUID of the Mapped Collection on the URL path.  So, the format is now `DELETE /api/core/items/<item:uuid>/mappedCollections/<collection:uuid>`

Please re-review and let me know what you think!

tdonohue commented:
@benbosman : While I understand your concerns about the `DELETE` syntax, I think we're somewhat forced into this syntax, not only by Spring Data REST best practices, but also via general REST best practices.  If you look at the [StackOverflow article](https://stackoverflow.com/a/299696/3750035) that @abollini referenced in his prior comments, there are a few key lines I'd like to stress in the answer:
> If a DELETE request includes an entity body, the body is ignored [...] 

And, in a comment below that:
> Please note that a lot of clients are also unable to send a DELETE with a body. This just burned me on Android.

So, it seems that this is not a question of "what does Spring do", but really a question of "what do REST Clients tend to support".  So, in my opinion, we're going to be forced into supporting `DELETE` requests that do *not* include information in the request body...but, I agree we'll need to clearly document this in the REST Contract.

tdonohue commented:
@benbosman : If you could give this another review, I'd appreciate it.  I've updated this REST Contract with the following:

* A note that `GET /api/core/items/<:uuid>/mappedCollections/<:collection_uuid>` is UNSUPPORTED
* A note that `DELETE  /api/core/items/<:uuid>/mappedCollections` is also UNSUPPORTED
* Modified README.md to note that `DELETE` cannot include a request body, and added examples for how to implement `DELETE` for associations (like mappedCollections)
* Minor formatting cleanup

tdonohue commented:
Thanks @benbosman and @abollini! Merging this as it's now at +2
